accel/tcg/log_instr.c

```diff
@@ -0,0 +1,513 @@
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+ * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+
+#include <stdint.h>
+
+#include "qemu/osdep.h"
+#include "qemu/log.h"
+#include "cpu.h"
+#include "exec/log.h"
+#include "exec/helper-proto.h"
+#include "exec/log_instr.h"
+#include "disas/disas.h"
+
+/*
+ * CHERI common instruction logging.
+ *
+ * This is the central implementation of the CPU_LOG_INSTR, CPU_LOG_CVTRACE
+ * and CPU_LOG_USER_ONLY methods.
+ * The same functions can be used by CHERI targets to append to the instruction
+ * log buffer. Once the instruction is fully processed, the target commits the
+ * log buffer and depending on the instruction operations and -dfilter options
+ * we either flush the buffer or drop it.
+ * A central design goal is to deterministically output the main side effect of
+ * the instruction (generally the destination register value), other side
+ * effects, such as multiple register updates, may be implemented by emitting
+ * multiple trace entries, but this is currently unimplemented.
+ * Another design goal is to allow tracing of arbitrary events via nops in
+ * both the text and binary format.
+ */
+
+#ifdef CONFIG_CHERI_LOG_INSTR
+
+#ifdef DEBUG_INSTR_LOGGING
+#define log_assert(x) assert((x))
+#else
+#define log_assert(x)
+#endif
+
+struct log_reginfo {
+    uint16_t flags;
+#define LRI_CAP_REG    1
+#define LRI_HOLDS_CAP  2
+
+    const char *name;
+    union {
+        target_ulong gpr;
+#ifdef TARGET_CHERI
+        cap_register_t cap;
+#endif
+    };
+};
+typedef struct log_reginfo log_reginfo_t;
+
+#define reginfo_is_cap(ri) (ri->flags & LRI_CAP_REG)
+#define reginfo_has_cap(ri) (reginfo_is_cap(ri) && (ri->flags & LRI_HOLDS_CAP))
+
+extern int cl_default_trace_format;
+
+static cpu_log_instr_info_t *cpu_get_log(CPUState *cpu)
+{
+    return &cpu->log_info;
+}
+
+static cpu_log_instr_info_t *target_cpu_get_log(CPUArchState *env)
+{
+    return &env_cpu(env)->log_info;
+}
+
+/*
+ * Emit textual trace entry to the log.
+ */
+static void emit_text_trace(CPUArchState *env, cpu_log_instr_info_t *log)
+{
+    int i;
+
+    /* Dump main instruction log */
+#if defined(TARGET_RISCV) && defined(CONFIG_RVFI_DII)
+    /*
+     * TODO(am2419): what to do with injected instructions?
```

Comment:
I wouldn't worry about that now. If it's not I can fix it later. We don't really need instruction logging then other than for debugging failures.